framework/DataObjects/DataSet.py

```diff
@@ -0,0 +1,1853 @@
+      if len(self._data[self.sampleTag])==0: #TODO what if it's just metadata?
+        self.raiseAWarning('Nothing to write!')
+        return
+      #first write the CSV
+      firstIndex = kwargs.get('firstIndex',0)
+      self._toCSV(fileName,start=firstIndex,**kwargs)
+      # then the metaxml
+      self._toCSVXML(fileName,**kwargs)
+    # TODO dask?
+    else:
+      self.raiseAnError(NotImplementedError,'Unrecognized write style: "{}"'.format(style))
+    if not self.hierarchical and 'RAVEN_isEnding' in self.getVars():
+      return len(self._data.where(self._data['RAVEN_isEnding']==True,drop=True)['RAVEN_isEnding'])
+    else:
+      return len(self) # so that other entities can track which realization we've written
+
+  ### INITIALIZATION ###
+  # These are the necessary functions to construct and initialize this data object
+  def __init__(self):#, in_vars, out_vars, meta_vars=None, dynamic=False, var_dims=None,cacheSize=100,prealloc=False):
+    """
+      Constructor.
+      @ In, None
+      @ Out, None
+    """
+    DataObject.__init__(self)
+    self.name      = 'DataSet'
+    self.type      = 'DataSet'
+    self.types     = None       # list of type objects, for each realization entry
+    self.printTag  = self.name
+    self.defaultDtype = object
+    self._scaleFactors = {}     # mean, sigma for data for matching purposes
+    self._alignedIndexes = {}   # dict {index:values} of indexes with aligned coordinates (so they are not in the collector, but here instead)
+
+  def _readMoreXML(self,xmlNode):
+    """
+      Initializes data object based on XML input
+      @ In, xmlNode, xml.etree.ElementTree.Element, input information
+      @ Out, None
+    """
+    inp = DataSet.getInputSpecification()()
+    inp.parseNode(xmlNode)
+    # let parent read first
+    DataObject._readMoreXML(self,inp)
+    # any additional custom reading below
+
+  ### BUIlTINS AND PROPERTIES ###
+  # These are special commands that RAVEN entities can use to interact with the data object
+  def __len__(self):
+    """
+      Overloads the len() operator.
+      @ In, None
+      @ Out, int, number of samples in this dataset
+    """
+    return self.size
+
+  @property
+  def vars(self):
+    """
+      Property to access all the pointwise variables being controlled by this data object.
+      As opposed to "self._orderedVars", returns the variables clustered by subset (inp, out, meta) instead of order added
+      @ In, None
+      @ Out, vars, list(str), variable names list
+    """
+    return self._inputs + self._outputs + self._metavars
+
+  @property
+  def size(self):
+    """
+      Property to access the amount of data in this data object.
+      @ In, None
+      @ Out, size, int, number of samples
+    """
+    s = 0 # counter for size
+    # from collector
+    s += self._collector.size if self._collector is not None else 0
+    # from data
+    s += len(self._data[self.sampleTag]) if self._data is not None else 0
```

Comment:
These would be so much simpler as proper if blocks, so they wouldn't need to else 0.